### ISSUE TITLE
Compiled Assets?

### ISSUE THREAD

**Issue by bengalih** [1nikolas/play-integrity-checker-app#1]
Are you able to create compiled assets for download?
This issue is hitting everyone right now and your checker is sorely needed to help people understand the issue.
Only found one compiled apk on XDA and it appears your v2 now shows more info:
https://github.com/kdrag0n/safetynet-fix/issues/203#issuecomment-1193137951

Hope it get to Google Play soon, but Google is notoriously slow!

**Comment by 1nikolas** [comment]
The app doesn't work because it's not published on Play Store 🥲

**Comment by bengalih** [comment]
Hey..so can you help me understand some of this a bit better.

1) I don't see anything in the docs which is stating that you can't request the labels of device integrity only when running from a server.  Is there not a raw response that comes back (like with SN) that shows the contents of the device_recognition_verdict?

How specifically (code-wise) do you request the "additional labels" in the device_recognition_verdict?

2) I am personally a bit confused by the two parts, the app and the server.  Where does the "server" piece need to run?  I assume you are running a server that the app is going through currently?  Meaning every time I use your app it has to go out to your server?  Is it the server that is evaluating the integrity or the device?   

IOW.  The device is always going to gather a certain amount of info (i.e. I pass check A,B,C, I fail check D).  It then has to pass that info to somewhere (your server?) which will decide what to do with that.  Does the server define what MEETS_STRONG_INTEGRITY is vs MEETS_BASIC_INTEGRITY?  Or are those qualifications the same for all devices across all apps?

If you can point me to the documentation or annotate some of your code points to clarify that would be helpful!

p.s.  If you want to close this issue and discuss in chat somewhere else, let me know.

**Comment by 1nikolas** [comment]
> where exactly is the evaluation of the device integrity being done

I assume Google can request info from your device remotely on demand. After all they get your location every few seconds so they can have accurate traffic on Google Maps, this wouldn't surprise me

**Comment by bengalih** [comment]
Right...so the PI API components that run on that machine basically say something like.
"We've been triggered to do a test."
So they gather info using some local code (like the local snet service for SafetyNet), including basicIntegrity and also the ctsProfile.

They then send this info to google which responds with device_recognition_verdict of either null or MEETS_DEVICE_INTEGRITY?

And, only if on Play store does it request MEETS_BASIC/STRONG_INTEGRITY and can return that to the app as well?

But, to be clear - in your current implementation the app that runs on my device is contacting your server over API, right?
Is your server doing all the decryption local at this point, or is it still sending anything to Google?

also, don't hold your breath with Google, it might take weeks from what I have heard. :/

**Comment by 1nikolas** [comment]
> So they gather info using some local code

Google Play Services probably does that

> They then send this info to google which responds with device_recognition_verdict of either null or MEETS_DEVICE_INTEGRITY?
> 
> And, only if on Play store does it request MEETS_BASIC/STRONG_INTEGRITY and can return that to the app as well?

Yes it can return from 0 to 3 flags

> in your current implementation the app that runs on my device is contacting your server over API, right?
>  Is your server doing all the decryption local at this point, or is it still sending anything to Google?

Well this was a demo app so I just put my Google Cloud keys and did the decryption by sending everything back to Google. I only did this because it was a demo, the Google Cloud project is going to get deleted.

> also, don't hold your breath with Google, it might take weeks from what I have heard. :/

Well let's hope they don't. On Google I found:
> This review process usually takes between 24 and 48 hours

Also I've published apps on Google Play before and I think it took around that time

**Comment by bengalih** [comment]
So further clarification.

What are the 4 flags (0,3) we are talking about in this case?
So if your app (apk) is going right to google, then currently the server/backend part of this project is not currently utilized at all in the process?

**Comment by 1nikolas** [comment]
> What are the 4 flags (0,3) we are talking about in this case?

I meant it can return nothing to 3 flags so it will have maximum 3 flags in the response

> So if your app (apk) is going right to google, then currently the server/backend part of this project is not currently utilized at all in the process?

Yes but it does in the new app

**Comment by bengalih** [comment]
So I just saw this picture in your documentation:

https://user-images.githubusercontent.com/30593419/180609045-fe0da305-24d9-4ffb-a44d-4294a759c787.png

Can you explain where that image comes from and how it is related to the one you posted here:
https://github.com/kdrag0n/safetynet-fix/issues/203#issuecomment-1193137951

?

I'm having some ideas as to how this is working together now, but don't have the backend setup to see myself.

thanks!

**Comment by 1nikolas** [comment]
> Can you explain where that image comes from

It's an option on Google Play Console (this is where you upload apps on Google Play)

> how it is related to the one you posted here:

This is a screenshot of the app

**Comment by bengalih** [comment]
Sorry I posted the wrong link.
 I meant how is this one:
ONE - https://user-images.githubusercontent.com/30593419/180609045-fe0da305-24d9-4ffb-a44d-4294a759c787.png

related to this one:
TWO - https://github.com/kdrag0n/safetynet-fix/issues/203#issuecomment-1192213589

Are they both in the Play Dev Console?  If so what sections are they under.
Reason I'm asking is that we were having discussion about the image in the comment...That it has 3 options:
- Don't Exclude
- Exclude only basic failures
- Exclude all device failures

But part of the confusion was that there is no differentiation between strong integrity in that selection.
However the specific text of the last one states:
"Exclude untrustworthy devices that don't meet device integrity"

If you recall, I was theorizing that here they meant "device integrity" = MEETS_STRONG_INTEGRITY because with only these options there would be no other way to filter on the two different types (3 if you count not to filter at all).

But image ONE looks like it is another section, so I'm wondering where that is and how they are related.

Based on this one I am thinking that "Device Integrity" and "MEETS_DEVICE_INTEGRITY" should never be used interchangeably and that in image ONE you basically tell it what you want "Device Integrity" to match by selecting the checkboxes.

If that is the case, then it gives image TWO a whole new meaning.  When it states
 "Exclude all device failures - Exclude untrustworthy devices that don't meet device integrity"
Do they mean the device integrity that you defined in ONE is what is being used as a criteria in TWO?

Do you see where I am getting at?  What do you think?  I don't have much insight because I don't have a dev account to look these things up on the console.

Also, I'm a bit perplexed by the choices in image ONE, but maybe your explanation can clear this up too.
There are two checkboxes under "Device Integrity" there:
MEETS_BASIC_INTEGRITY
MEETS_STRONG_INTEGRITY

So I understand if you want the app to verify that the device at least passes BASIC you would select the BASIC check box.
If you want to ensure that it passes STRONG, you would check the STRONG check box.
However, doesn't passing STRONG assume that BASIC has to be passed as well? 
I'm pretty sure we are all agreeing it does.
If that is the case, checking the STRONG box should automatically check the BASIC.   Does it?
If it doesn't that could be interesting (or just bad UI design?)
It would seem this should be a checkbox, but a list box and the three choices should be None, Basic, or Strong, with the latter inferring Basic is included.

Does that make sense?  Just trying to figure it out based on what they've given us.  I'd like to give them the credit that the documentation and the UI means what it says, but it doesn't always seem to say the same thing.

**Comment by 1nikolas** [comment]
This option
 [](https://user-images.githubusercontent.com/30593419/180609045-fe0da305-24d9-4ffb-a44d-4294a759c787.png)

Is for toggling strong and basic flags for the api response which gets to the app (and server backend)

[](https://user-images.githubusercontent.com/30593419/180323882-dfa68e9d-e077-4711-b927-f1ff1a29b45f.png)

This is for (based on the flags) not let devices that don't pass these flags to download it from playstore. The checks are done on Play Store (Google Play calls the api, gets response, etc)

**Comment by bengalih** [comment]
So do you agree with my statement then?
Does "device integrity" in the statement "Exclude untrustworthy device that don't meet device integrity" mean the options configured in "Edit device integrity labels" ?
IOW the exclusions from option 3 is dependent on what it set in the Labels option?

**Comment by 1nikolas** [comment]
> So do you agree with my statement then?
> Does "device integrity" in the statement "Exclude untrustworthy device that don't meet device integrity" mean the options configured in "Edit device integrity labels" ?
> IOW the exclusions from option 3 is dependent on what it set in the Labels option?

I have no idea how the "disallow download from Play Store" thing works. It's all internally done by Google. For the other one, we'll find out

**Comment by 1nikolas** [comment]
It's now available on Play Store

https://play.google.com/store/apps/details?id=gr.nikolasspyr.integritycheck
